### PR TITLE
MGMT-7741: New validation: TANG server connectivity check 

### PR DESCRIPTION
**Pull request by nmagnezi**
## List all the issues related to this PR

- [x] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [x] Automation (CI, tools, etc)
- [x] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Assignees

/cc @
/cc @

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by filanov** [comment]
@tsorya is the format of the command aligned with the security format?

File: internal/host/hostcommands/tang_connectivity_check_cmd.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+	if cluster.DiskEncryption == nil ||
+		swag.StringValue(cluster.DiskEncryption.EnableOn) == models.DiskEncryptionEnableOnNone ||
+		swag.StringValue(cluster.DiskEncryption.Mode) == models.DiskEncryptionModeTpmv2 ||
+		!hostutil.IsDiskEncryptionEnabledForRole(*cluster.DiskEncryption, common.GetEffectiveRole(host)) {
+		c.log.Debugf("skipping tangConnectivityCheck for host %s, DiskEncryption config do not require validation here", host.ID.String())
+		return []*models.Step{}, nil
+	}
+
+	step := &models.Step{
```
Comment (tsorya) [inline internal/host/hostcommands/tang_connectivity_check_cmd.go:49]: all this should move to agent

**Comment by tsorya** [comment]
@nmagnezi you should send only arguments that are dynamic, all other logic should be inside agent. You should add validation per each string argument that you send (if it is possible)

**Comment by tsorya** [comment]
> @tsorya is the format of the command aligned with the security format?

@filanov you are totally right, thank you for pointing to it.

File: internal/host/hostcommands/tang_connectivity_check_cmd.go (hunk 1 above)
Comment (nmagnezi) [reply internal/host/hostcommands/tang_connectivity_check_cmd.go:44]: Done.
Now I'm only passing `cluster.DiskEncryption.TangServers` as a param.
Implementing the agent side in https://github.com/openshift/assisted-installer-agent/pull/341 (in progress)

File: internal/host/hostcommands/tang_connectivity_check_cmd.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+	if err := c.db.First(&cluster, "id = ?", host.ClusterID).Error; err != nil {
+		c.log.WithError(err).Errorf("failed to fetch cluster %s", host.ClusterID)
+		return nil, err
+	}
+
+	if cluster.DiskEncryption == nil ||
+		swag.StringValue(cluster.DiskEncryption.EnableOn) == models.DiskEncryptionEnableOnNone ||
+		swag.StringValue(cluster.DiskEncryption.Mode) == models.DiskEncryptionModeTpmv2 ||
+		!hostutil.IsDiskEncryptionEnabledForRole(*cluster.DiskEncryption, common.GetEffectiveRole(host)) {
```
Comment (danielerez) [inline internal/host/hostcommands/tang_connectivity_check_cmd.go:39]: It's a bit hard to follow this condition:) can you maybe split this to vars or add some comment. just to make it clearer.

File: internal/cluster/validations/validations.go [hunk 3]
```diff
@@ -501,7 +502,7 @@ func ValidateDiskEncryptionParams(diskEncryptionParams *models.DiskEncryption, D
 		if diskEncryptionParams.TangServers == "" {
 			return errors.New("Setting Tang mode but tang_servers isn't set")
 		}
-		tangServers, err := common.UnmarshalTangServers(diskEncryptionParams.TangServers)
```
Comment (danielerez) [inline internal/cluster/validations/validations.go:504]: most unmarshal helpers are actually on common? any reason to move it?

File: swagger.yaml [hunk 4]
```diff
@@ -5461,6 +5465,44 @@ definitions:
...
+        type: string
+        description: JSON-formatted string containing additional information regarding tang's configuration
+
+  tang_connectivity_response:
+    type: object
+    properties:
+      is_success:
+        type: boolean
+        description: Tang check result.
+      tang_server_response:
```
Comment (danielerez) [inline swagger.yaml:5483]: are we using the tang_server_response somewhere? iiuc, we only check whether tang_connectivity_response succeeded for now?

Comment (nmagnezi) [reply swagger.yaml:5483]: I use it in the agent code to make sure we get a valid response from Tang servers: see [TangRequest()](https://github.com/openshift/assisted-installer-agent/pull/341/files#diff-d1bab7abd87cd22a53866c28c672fb7386f432d9c44353543f3e3a4021afd987R74)

File: internal/cluster/validations/validations.go (hunk 3 above)
Comment (nmagnezi) [reply internal/cluster/validations/validations.go:504]: I moved it out of `/internal` because I needed this function for the agent code as well.
See [CheckTangConnectivity()](https://github.com/openshift/assisted-installer-agent/pull/341/files#diff-d1bab7abd87cd22a53866c28c672fb7386f432d9c44353543f3e3a4021afd987R74) in https://github.com/openshift/assisted-installer-agent/pull/341

File: internal/host/hostcommands/tang_connectivity_check_cmd.go (hunk 2 above)
Comment (nmagnezi) [reply internal/host/hostcommands/tang_connectivity_check_cmd.go:39]: Ack.
How about:
```
	// Skip tangConnectivityCheck for cases where:
	// 1. DiskEncryption not set or not enabled.
	// 2. DiskEncryption mode is not tang based.
	// 3. DiskEncryption is not enabled, for the host role.
```
The last condition is the same helper function we currently use in `diskEncryptionRequirementsSatisfied` validation. In My cast I'm using it with a `not`.

Comment (danielerez) [reply internal/host/hostcommands/tang_connectivity_check_cmd.go:39]: Yeah, should be clearer.

File: swagger.yaml (hunk 4 above)
Comment (danielerez) [reply swagger.yaml:5483]: oh ok, so TangServerSignatures is also mandatory for the response to be considered valid?

Comment (nmagnezi) [reply swagger.yaml:5483]: Yes, this is an application-level validation (so it is not enough for the server to respond with 200 OK), with the specific thumbprint configured by the user.

WIth the agent code, we will Unmarshal the response we get from Tang, to make sure we get keys (Array) that are specific to the used thumbprint. This is to make sure as much as we can that upon reboot, things would work.

File: internal/host/hostcommands/tang_connectivity_check_cmd.go (hunk 2 above)
Comment (nmagnezi) [reply internal/host/hostcommands/tang_connectivity_check_cmd.go:43]: Done.

File: internal/host/hostutil/test_utils.go [hunk 5]
```diff
@@ -163,6 +163,41 @@ func GenerateTestAPIVIpConnectivity(ignition string) string {
 	Expect(err).To(Not(HaveOccurred()))
 	return string(bytes)
 }
+func getTangResponse(url string) models.TangServerResponse {
```
Comment (ybettan) [inline internal/host/hostutil/test_utils.go:166]: This code is only for testing purpose?

File: internal/host/validator.go [hunk 6]
```diff
@@ -414,6 +401,19 @@ func (v *validator) getDiskEncryptionForDay2(host *models.Host) (*types.Luks, er
 	return &config.Storage.Luks[0], nil
 }
 
+func (v *validator) areTangServersReachable(c *validationContext) ValidationStatus {
+	if c.host.TangConnectivity == "" {
```
Comment (ybettan) [inline internal/host/validator.go:405]: How do we differ this case from the case in which an old agent that doesn't run the tang validation doesn't send the info?

The fact that this function is missing a flow in which we return `ValidationSuccessSuppressOutput` make me suspicious :)

File: internal/host/hostutil/test_utils.go (hunk 5 above)
Comment (nmagnezi) [reply internal/host/hostutil/test_utils.go:166]: yes

File: internal/bminventory/inventory.go [hunk 7]
```diff
@@ -2788,6 +2788,21 @@ func (b *bareMetalInventory) handleReplyError(params installer.V2PostStepReplyPa
...
 		}
 		return b.processDiskSpeedCheckResponse(ctx, h, stepReply, exitCode)
+
+	case models.StepTypeTangConnectivityCheck:
+		// Older agents have no action for tang-connectivity-check.
+		// The missing action result will not fail host validations to
+		// keep backward compatibility with older agents who did not include tang-connectivity-check.
+		missingTangConnectivityActionError := fmt.Sprintf("failed to find action for step type %s", models.StepTypeTangConnectivityCheck)
+
+		if params.Reply.Error == missingTangConnectivityActionError {
```
Comment (danielerez) [inline internal/bminventory/inventory.go:2798]: maybe safer to check if the error just contains 'models.StepTypeTangConnectivityCheck' (in case the exact string changes in agent)

File: internal/host/validator.go (hunk 6 above)
Comment (nmagnezi) [reply internal/host/validator.go:406]: Thanks for the comment!
I moved the code that handled "no such action" to `areTangServersReachable()` and used `ValidationSuccessSuppressOutput`.

I re-tested e2e both positive and negative cases and it looks good. Printing this to the log:
```go
time="2022-07-07T11:32:53Z" level=warning msg="host 44ac8510-cb48-4481-9284-d8854033da48 replied to StepType: tang-connectivity-check with: failed to find action for step type tang-connectivity-check. Validation will pass to keep backward compatibility with discovery agent version: latest" func="github.com/openshift/assisted-service/internal/host.(*validator).areTangServersReachable" file="/go/src/github.com/openshift/origin/internal/host/validator.go:413" pkg=host-state
```

**Comment by ybettan** [comment]
/lgtm
/hold
Unhold when you want to merge :)
